### ISSUE TITLE
NotInitializedError: The proper initialization has not occurred. The mongodb database is missing account_controls, accounts, pub_keys, transaction_traces, transactions, blocks collections.

### ISSUE THREAD

**Issue by marinrusu1997** [EOSIO/demux-js-eos#55]
I can't connect to mongo database. I always get this error, even if database has these collections. Also, I observed that if I create a new database adding say only transaction_traces, I'm getting error that this collection is missing. Here's my code and error:

![screenshot from 2019-03-01 01-35-02](https://user-images.githubusercontent.com/29524251/53606101-3f1bee00-3bc2-11e9-9ed1-b202aa8c249b.png)

![screenshot from 2019-03-01 01-32-59](https://user-images.githubusercontent.com/29524251/53606023-fbc17f80-3bc1-11e9-9134-aad5112cf98d.png)

![screenshot from 2019-03-01 01-34-11](https://user-images.githubusercontent.com/29524251/53606073-24497980-3bc2-11e9-89b1-8e64ef5f1d15.png)

**Comment by jeffreyssmith2nd** [comment]
Does this happen if you've had the mongodb database (and nodeos) running for awhile and then start up demux?

**Comment by marinrusu1997** [comment]
> Does this happen if you've had the mongodb database (and nodeos) running for awhile and then start up demux?

Yes, this happens always. I switched now to demux 3.0.0 and demux-eos 2.0.0 and it seems to work fine, but with 4.0.0 versions it doesn't

**Comment by jeffreyssmith2nd** [comment]
Okay thanks for the report, we'll take a look at it.

**Comment by flux627** [comment]
Did you update the signature of `MongoActionReader` when upgrading to 4.0.0 [as per the release notes](https://github.com/EOSIO/demux-js-eos/releases/tag/4.0.0)? It now requires a [`MongoActionReaderOptions` object](https://github.com/EOSIO/demux-js-eos/blob/develop/src/interfaces.ts#L7-L10) (which extends the [`ActionReaderOptions` interface](https://github.com/EOSIO/demux-js/blob/develop/src/interfaces.ts#L1-L15)).

**Comment by marinrusu1997** [comment]
> Did you update the signature of `MongoActionReader` when upgrading to 4.0.0 [as per the release notes](https://github.com/EOSIO/demux-js-eos/releases/tag/4.0.0)? It now requires a [`MongoActionReaderOptions` object](https://github.com/EOSIO/demux-js-eos/blob/develop/src/interfaces.ts#L7-L10) (which extends the [`ActionReaderOptions` interface](https://github.com/EOSIO/demux-js/blob/develop/src/interfaces.ts#L1-L15)).

this happens even if I use constructor with default parametters, they are the same as my db instance

**Comment by flux627** [comment]
It looks like this is the problem:
https://github.com/EOSIO/demux-js-eos/blob/develop/src/mongo/MongoActionReader.ts#L111-L116

We should be checking if the needed collections are a subset of the existing, not the other way around. You can feel free to make a PR, if not I'll get around to it sometime next week.

**Comment by marinrusu1997** [comment]
> It looks like this is the problem:
> https://github.com/EOSIO/demux-js-eos/blob/develop/src/mongo/MongoActionReader.ts#L111-L116
> 
> We should be checking if the needed collections are a subset of the existing, not the other way around. You can feel free to make a PR, if not I'll get around to it sometime next week.

You were right, I changed code a little bit. Here's my workaround:
setup() method
![image](https://user-images.githubusercontent.com/29524251/53673765-3c88c980-3c92-11e9-83fe-4d431537f77e.png)
constructor
![image](https://user-images.githubusercontent.com/29524251/53673815-7954c080-3c92-11e9-9d78-018fb5f58a03.png)

And as you said, for 4.0.0 version I had to pass as parametter an object which implements both Mongo and Abstract Reader Options:
![image](https://user-images.githubusercontent.com/29524251/53674413-74920b80-3c96-11e9-91dd-3e637927569f.png)

But it seems that it's a little bug gere. I mean sometimes even if i pass -1, he is not reading from the tail, but from the beginning. I don't know what's the problem, maybe because I'm using this with MassiveActionHandler, but I'm not sure
